# **New York City Yellow Taxi Data**

## Objective
In this case study you will be learning exploratory data analysis (EDA) with the help of a dataset on yellow taxi rides in New York City. This will enable you to understand why EDA is an important step in the process of data science and machine learning.

## **Problem Statement**
As an analyst at an upcoming taxi operation in NYC, you are tasked to use the 2023 taxi trip data to uncover insights that could help optimise taxi operations. The goal is to analyse patterns in the data that can inform strategic decisions to improve service efficiency, maximise revenue, and enhance passenger experience.

## Tasks
You need to perform the following steps for successfully completing this assignment:
1. Data Loading
2. Data Cleaning
3. Exploratory Analysis: Bivariate and Multivariate
4. Creating Visualisations to Support the Analysis
5. Deriving Insights and Stating Conclusions

---

**NOTE:** The marks given along with headings and sub-headings are cumulative marks for those particular headings/sub-headings.<br>

The actual marks for each task are specified within the tasks themselves.

For example, marks given with heading *2* or sub-heading *2.1* are the cumulative marks, for your reference only. <br>

The marks you will receive for completing tasks are given with the tasks.

Suppose the marks for two tasks are: 3 marks for 2.1.1 and 2 marks for 3.2.2, or
* 2.1.1 [3 marks]
* 3.2.2 [2 marks]

then, you will earn 3 marks for completing task 2.1.1 and 2 marks for completing task 3.2.2.


---

## Data Understanding
The yellow taxi trip records include fields capturing pick-up and drop-off dates/times, pick-up and drop-off locations, trip distances, itemized fares, rate types, payment types, and driver-reported passenger counts.

The data is stored in Parquet format (*.parquet*). The dataset is from 2009 to 2024. However, for this assignment, we will only be using the data from 2023.

The data for each month is present in a different parquet file. You will get twelve files for each of the months in 2023.

The data was collected and provided to the NYC Taxi and Limousine Commission (TLC) by technology providers like vendors and taxi hailing apps. <br>

You can find the link to the TLC trip records page here: https://www.nyc.gov/site/tlc/about/tlc-trip-record-data.page

###  Data Description
You can find the data description here: [Data Dictionary](https://www.nyc.gov/assets/tlc/downloads/pdf/data_dictionary_trip_records_yellow.pdf)

**Trip Records**



|Field Name       |description |
|:----------------|:-----------|
| VendorID | A code indicating the TPEP provider that provided the record. <br> 1= Creative Mobile Technologies, LLC; <br> 2= VeriFone Inc. |
| tpep_pickup_datetime | The date and time when the meter was engaged.  |
| tpep_dropoff_datetime | The date and time when the meter was disengaged.   |
| Passenger_count | The number of passengers in the vehicle. <br> This is a driver-entered value. |
| Trip_distance | The elapsed trip distance in miles reported by the taximeter. |
| PULocationID | TLC Taxi Zone in which the taximeter was engaged |
| DOLocationID | TLC Taxi Zone in which the taximeter was disengaged |
|RateCodeID |The final rate code in effect at the end of the trip.<br> 1 = Standard rate <br> 2 = JFK <br> 3 = Newark <br>4 = Nassau or Westchester <br>5 = Negotiated fare <br>6 = Group ride |
|Store_and_fwd_flag |This flag indicates whether the trip record was held in vehicle memory before sending to the vendor, aka “store and forward,” because the vehicle did not have a connection to the server.  <br>Y= store and forward trip <br>N= not a store and forward trip |
|Payment_type| A numeric code signifying how the passenger paid for the trip. <br> 1 = Credit card <br>2 = Cash <br>3 = No charge <br>4 = Dispute <br>5 = Unknown <br>6 = Voided trip |
|Fare_amount| The time-and-distance fare calculated by the meter. <br>Extra Miscellaneous extras and surcharges.  Currently, this only includes the 0.50 and 1 USD rush hour and overnight charges. |
|MTA_tax |0.50 USD MTA tax that is automatically triggered based on the metered rate in use. |
|Improvement_surcharge | 0.30 USD improvement surcharge assessed trips at the flag drop. The improvement surcharge began being levied in 2015. |
|Tip_amount |Tip amount – This field is automatically populated for credit card tips. Cash tips are not included. |
| Tolls_amount | Total amount of all tolls paid in trip.  |
| total_amount | The total amount charged to passengers. Does not include cash tips. |
|Congestion_Surcharge |Total amount collected in trip for NYS congestion surcharge. |
| Airport_fee | 1.25 USD for pick up only at LaGuardia and John F. Kennedy Airports|

Although the amounts of extra charges and taxes applied are specified in the data dictionary, you will see that some cases have different values of these charges in the actual data.

**Taxi Zones**

Each of the trip records contains a field corresponding to the location of the pickup or drop-off of the trip, populated by numbers ranging from 1-263.

These numbers correspond to taxi zones, which may be downloaded as a table or map/shapefile and matched to the trip records using a join.

This is covered in more detail in later sections.

---

## **1** Data Preparation

<font color = red>[5 marks]</font> <br>

### Import Libraries

In [2]:
# Import warnings

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
# Import the libraries you will be using for analysis



In [3]:
# Recommended versions
# numpy version: 1.26.4
# pandas version: 2.2.2
# matplotlib version: 3.10.0
# seaborn version: 0.13.2

# Check versions
print("numpy version:", np.__version__)
print("pandas version:", pd.__version__)
print("matplotlib version:", plt.matplotlib.__version__)
print("seaborn version:", sns.__version__)

numpy version: 1.26.4
pandas version: 2.2.2
matplotlib version: 3.10.0
seaborn version: 0.13.2


### **1.1** Load the dataset
<font color = red>[5 marks]</font> <br>

You will see twelve files, one for each month.

To read parquet files with Pandas, you have to follow a similar syntax as that for CSV files.

`df = pd.read_parquet('file.parquet')`

In [5]:
# Try loading one file

# df = pd.read_parquet('2023-1.parquet')
# df.info()

df=pd.read_parquet('yellow_tripdata_2023-01.parquet')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3066766 entries, 0 to 3066765
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee           

How many rows are there? Do you think handling such a large number of rows is computationally feasible when we have to combine the data for all twelve months into one?

To handle this, we need to sample a fraction of data from each of the files. How to go about that? Think of a way to select only some portion of the data from each month's file that accurately represents the trends.

#### Sampling the Data
> One way is to take a small percentage of entries for pickup in every hour of a date. So, for all the days in a month, we can iterate through the hours and select 5% values randomly from those. Use `tpep_pickup_datetime` for this. Separate date and hour from the datetime values and then for each date, select some fraction of trips for each of the 24 hours.

To sample data, you can use the `sample()` method. Follow this syntax:

```Python
# sampled_data is an empty DF to keep appending sampled data of each hour
# hour_data is the DF of entries for an hour 'X' on a date 'Y'

sample = hour_data.sample(frac = 0.05, random_state = 42)
# sample 0.05 of the hour_data
# random_state is just a seed for sampling, you can define it yourself

sampled_data = pd.concat([sampled_data, sample]) # adding data for this hour to the DF
```

This *sampled_data* will contain 5% values selected at random from each hour.

Note that the code given above is only the part that will be used for sampling and not the complete code required for sampling and combining the data files.

Keep in mind that you sample by date AND hour, not just hour. (Why?)

In [15]:
sampled_data_def1 = df.sample(frac=0.0075, random_state=42)
sampled_data_def1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 21855 entries, 1288541 to 1408601
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   VendorID               21855 non-null  int32         
 1   tpep_pickup_datetime   21855 non-null  datetime64[us]
 2   tpep_dropoff_datetime  21855 non-null  datetime64[us]
 3   passenger_count        21295 non-null  float64       
 4   trip_distance          21855 non-null  float64       
 5   RatecodeID             21295 non-null  float64       
 6   store_and_fwd_flag     21295 non-null  object        
 7   PULocationID           21855 non-null  int32         
 8   DOLocationID           21855 non-null  int32         
 9   payment_type           21855 non-null  int64         
 10  fare_amount            21855 non-null  float64       
 11  extra                  21855 non-null  float64       
 12  mta_tax                21855 non-null  float64       
 13

In [19]:
df2=pd.read_parquet('yellow_tripdata_2023-02.parquet')
sampled_data_def2 = df2.sample(frac=0.0075, random_state=42)
sampled_data_def2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 21855 entries, 1288541 to 1408601
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   VendorID               21855 non-null  int32         
 1   tpep_pickup_datetime   21855 non-null  datetime64[us]
 2   tpep_dropoff_datetime  21855 non-null  datetime64[us]
 3   passenger_count        21295 non-null  float64       
 4   trip_distance          21855 non-null  float64       
 5   RatecodeID             21295 non-null  float64       
 6   store_and_fwd_flag     21295 non-null  object        
 7   PULocationID           21855 non-null  int32         
 8   DOLocationID           21855 non-null  int32         
 9   payment_type           21855 non-null  int64         
 10  fare_amount            21855 non-null  float64       
 11  extra                  21855 non-null  float64       
 12  mta_tax                21855 non-null  float64       
 13

In [20]:
df3=pd.read_parquet('yellow_tripdata_2023-03.parquet')
sampled_data_def3 = df3.sample(frac=0.0075, random_state=42)
sampled_data_def3.info()

<class 'pandas.core.frame.DataFrame'>
Index: 25528 entries, 2904679 to 2653750
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   VendorID               25528 non-null  int32         
 1   tpep_pickup_datetime   25528 non-null  datetime64[us]
 2   tpep_dropoff_datetime  25528 non-null  datetime64[us]
 3   passenger_count        24865 non-null  float64       
 4   trip_distance          25528 non-null  float64       
 5   RatecodeID             24865 non-null  float64       
 6   store_and_fwd_flag     24865 non-null  object        
 7   PULocationID           25528 non-null  int32         
 8   DOLocationID           25528 non-null  int32         
 9   payment_type           25528 non-null  int64         
 10  fare_amount            25528 non-null  float64       
 11  extra                  25528 non-null  float64       
 12  mta_tax                25528 non-null  float64       
 13

In [21]:
df4=pd.read_parquet('yellow_tripdata_2023-04.parquet')
df5=pd.read_parquet('yellow_tripdata_2023-05.parquet')
df6=pd.read_parquet('yellow_tripdata_2023-06.parquet')
df7=pd.read_parquet('yellow_tripdata_2023-07.parquet')
df8=pd.read_parquet('yellow_tripdata_2023-08.parquet')
df9=pd.read_parquet('yellow_tripdata_2023-09.parquet')
df10=pd.read_parquet('yellow_tripdata_2023-10.parquet')
df11=pd.read_parquet('yellow_tripdata_2023-11.parquet')
df12=pd.read_parquet('yellow_tripdata_2023-12.parquet')
sampled_data_def4 = df4.sample(frac=0.0075, random_state=42)
sampled_data_def5 = df5.sample(frac=0.0075, random_state=42)
sampled_data_def6 = df6.sample(frac=0.0075, random_state=42)
sampled_data_def7 = df7.sample(frac=0.0075, random_state=42)
sampled_data_def8 = df8.sample(frac=0.0075, random_state=42)
sampled_data_def9 = df9.sample(frac=0.0075, random_state=42)
sampled_data_def10 = df10.sample(frac=0.0075, random_state=42)
sampled_data_def11 = df11.sample(frac=0.0075, random_state=42)
sampled_data_def12 = df12.sample(frac=0.0075, random_state=42)


In [41]:
df_list = [sampled_data_def1,sampled_data_def2,sampled_data_def3,sampled_data_def4,
           sampled_data_def5,sampled_data_def6,sampled_data_def7,sampled_data_def8,
           sampled_data_def9,sampled_data_def10,sampled_data_def11,sampled_data_def12]

In [53]:
com_df = pd.concat(df_list, ignore_index=True)
com_df.to_parquet("merged_output.parquet", index=False)
com_df.to_csv("merged_output.csv", index=False)

In [50]:
com_df

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0              2  2023-02-13 17:11:02   2023-02-13 17:16:46              1.0   
1              2  2023-02-05 18:39:05   2023-02-05 19:10:29              2.0   
2              2  2023-02-21 10:15:06   2023-02-21 10:51:25              1.0   
3              2  2023-02-24 18:40:20   2023-02-24 18:55:07              1.0   
4              2  2023-02-11 08:55:21   2023-02-11 09:16:10              1.0   
...          ...                  ...                   ...              ...   
286175         2  2023-12-12 09:40:32   2023-12-12 09:44:36              1.0   
286176         2  2023-12-30 12:50:08   2023-12-30 13:21:16              1.0   
286177         1  2023-12-28 12:57:03   2023-12-28 13:12:19              2.0   
286178         2  2023-12-10 15:37:47   2023-12-10 15:49:04              1.0   
286179         2  2023-12-14 22:42:23   2023-12-14 22:48:38              1.0   

        trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                0.94         1.0                  N           141   
1               10.39         1.0                  N           138   
2               10.26         1.0                  N           138   
3                1.53         1.0                  N           234   
4               16.71         2.0                  N           137   
...               ...         ...                ...           ...   
286175           1.01         1.0                  N           238   
286176           6.85         1.0                  N           231   
286177           2.50         1.0                  N            41   
286178           2.22         1.0                  N           163   
286179           1.11         1.0                  N           164   

        DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                262             1          7.9    2.5      0.5        2.88   
1                 61             1         45.0    5.0      0.5       10.00   
2                163             1         47.8    5.0      0.5        3.00   
3                230             1         14.2    2.5      0.5        4.14   
4                132             1         70.0    0.0      0.5       24.16   
...              ...           ...          ...    ...      ...         ...   
286175           142             1          7.2    0.0      0.5        2.80   
286176            43             1         36.6    0.0      0.5        8.12   
286177           239             1         16.3    2.5      0.5        4.05   
286178           236             1         14.2    0.0      0.5        3.64   
286179           234             1          8.6    1.0      0.5        2.00   

        tolls_amount  improvement_surcharge  total_amount  \
0               0.00                    1.0         17.28   
1               0.00                    1.0         62.75   
2               6.55                    1.0         67.60   
3               0.00                    1.0         24.84   
4               6.55                    1.0        104.71   
...              ...                    ...           ...   
286175          0.00                    1.0         14.00   
286176          0.00                    1.0         48.72   
286177          0.00                    1.0         24.35   
286178          0.00                    1.0         21.84   
286179          0.00                    1.0         15.60   

        congestion_surcharge  Airport_fee  
0                        2.5         0.00  
1                        0.0         1.25  
2                        2.5         1.25  
3                        2.5         0.00  
4                        2.5         0.00  
...                      ...          ...  
286175                   2.5         0.00  
286176                   2.5         0.00  
286177                   2.5         0.00  
286178                   2.5         0.00  
286179                   2.5         0.0

**1.1.1** <font color = red>[5 marks]</font> <br>
Figure out how to sample and combine the files.

In [ ]:
com_df

**Note:** It is not mandatory to use the method specified above. While sampling, you only need to make sure that your sampled data represents the overall data of all the months accurately.

In [ ]:
# Sample the data
# It is recommmended to not load all the files at once to avoid memory overload

In [ ]:
# from google.colab import drive
# drive.mount('/content/drive')

In [ ]:
# Take a small percentage of entries from each hour of every date.
# Iterating through the monthly data:
#   read a month file -> day -> hour: append sampled data -> move to next hour -> move to next day after 24 hours -> move to next month file
# Create a single dataframe for the year combining all the monthly data

# Select the folder having data files
import os

# Select the folder having data files
os.chdir('/content/Assignments/EDA/data_NYC_Taxi/trip_records')

# Create a list of all the twelve files to read
file_list = os.listdir()

# initialise an empty dataframe
df = pd.DataFrame()


# iterate through the list of files and sample one by one:
for file_name in file_list:
    try:
        # file path for the current file
        file_path = os.path.join(os.getcwd(), file_name)

        # Reading the current file



        # We will store the sampled data for the current date in this df by appending the sampled data from each hour to this
        # After completing iteration through each date, we will append this data to the final dataframe.
        sampled_data = pd.DataFrame()

        # Loop through dates and then loop through every hour of each date

            # Iterate through each hour of the selected date

                # Sample 5% of the hourly data randomly

                # add data of this hour to the dataframe

        # Concatenate the sampled data of all the dates to a single dataframe
        df = # we initialised this empty DF earlier

    except Exception as e:
        print(f"Error reading file {file_name}: {e}")

After combining the data files into one DataFrame, convert the new DataFrame to a CSV or parquet file and store it to use directly.

Ideally, you can try keeping the total entries to around 250,000 to 300,000.

In [ ]:
# Store the df in csv/parquet
# df.to_parquet('')

## **2** Data Cleaning
<font color = red>[30 marks]</font> <br>

Now we can load the new data directly.

In [22]:
# Load the new data file



In [54]:
# df.head()
com_df.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2023-02-13 17:11:02   2023-02-13 17:16:46              1.0   
1         2  2023-02-05 18:39:05   2023-02-05 19:10:29              2.0   
2         2  2023-02-21 10:15:06   2023-02-21 10:51:25              1.0   
3         2  2023-02-24 18:40:20   2023-02-24 18:55:07              1.0   
4         2  2023-02-11 08:55:21   2023-02-11 09:16:10              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           0.94         1.0                  N           141           262   
1          10.39         1.0                  N           138            61   
2          10.26         1.0                  N           138           163   
3           1.53         1.0                  N           234           230   
4          16.71         2.0                  N           137           132   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             1          7.9    2.5      0.5        2.88          0.00   
1             1         45.0    5.0      0.5       10.00          0.00   
2             1         47.8    5.0      0.5        3.00          6.55   
3             1         14.2    2.5      0.5        4.14          0.00   
4             1         70.0    0.0      0.5       24.16          6.55   

   improvement_surcharge  total_amount  congestion_surcharge  Airport_fee  
0                    1.0         17.28                   2.5         0.00  
1                    1.0         62.75                   0.0         1.25  
2                    1.0         67.60                   2.5         1.25  
3                    1.0         24.84                   2.5         0.00  
4                    1.0        104.71                   2.5         0.00

In [56]:
# df.info()
com_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 286180 entries, 0 to 286179
Data columns (total 19 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   VendorID               286180 non-null  int32         
 1   tpep_pickup_datetime   286180 non-null  datetime64[us]
 2   tpep_dropoff_datetime  286180 non-null  datetime64[us]
 3   passenger_count        276296 non-null  float64       
 4   trip_distance          286180 non-null  float64       
 5   RatecodeID             276296 non-null  float64       
 6   store_and_fwd_flag     276296 non-null  object        
 7   PULocationID           286180 non-null  int32         
 8   DOLocationID           286180 non-null  int32         
 9   payment_type           286180 non-null  int64         
 10  fare_amount            286180 non-null  float64       
 11  extra                  286180 non-null  float64       
 12  mta_tax                286180 non-null  floa

#### **2.1** Fixing Columns
<font color = red>[10 marks]</font> <br>

Fix/drop any columns as you seem necessary in the below sections

In [57]:
com_df.tail()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
286175         2  2023-12-12 09:40:32   2023-12-12 09:44:36              1.0   
286176         2  2023-12-30 12:50:08   2023-12-30 13:21:16              1.0   
286177         1  2023-12-28 12:57:03   2023-12-28 13:12:19              2.0   
286178         2  2023-12-10 15:37:47   2023-12-10 15:49:04              1.0   
286179         2  2023-12-14 22:42:23   2023-12-14 22:48:38              1.0   

        trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
286175           1.01         1.0                  N           238   
286176           6.85         1.0                  N           231   
286177           2.50         1.0                  N            41   
286178           2.22         1.0                  N           163   
286179           1.11         1.0                  N           164   

        DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
286175           142             1          7.2    0.0      0.5        2.80   
286176            43             1         36.6    0.0      0.5        8.12   
286177           239             1         16.3    2.5      0.5        4.05   
286178           236             1         14.2    0.0      0.5        3.64   
286179           234             1          8.6    1.0      0.5        2.00   

        tolls_amount  improvement_surcharge  total_amount  \
286175           0.0                    1.0         14.00   
286176           0.0                    1.0         48.72   
286177           0.0                    1.0         24.35   
286178           0.0                    1.0         21.84   
286179           0.0                    1.0         15.60   

        congestion_surcharge  Airport_fee  
286175                   2.5          0.0  
286176                   2.5          0.0  
286177                   2.5          0.0  
286178                   2.5          0.0  
286179                   2.5          0.0

**2.1.1** <font color = red>[2 marks]</font> <br>

Fix the index and drop unnecessary columns

In [25]:
# Fix the index and drop any columns that are not needed

      ## Fixed index and time being i will not drop any columns

**2.1.2** <font color = red>[3 marks]</font> <br>
There are two airport fee columns. This is possibly an error in naming columns. Let's see whether these can be combined into a single column.

In [26]:
# Combine the two airport fee columns

      ## In my file only one column appeared

**2.1.3** <font color = red>[5 marks]</font> <br>
Fix columns with negative (monetary) values

In [61]:
# check where values of fare amount are negative

(com_df['fare_amount']<0).value_counts()

fare_amount
False    283259
True       2921
Name: count, dtype: int64

In [63]:
com_df['fare_amount'] = com_df['fare_amount'].apply(lambda x: 0 if x < 0 else x)
(com_df['fare_amount']<0).value_counts()


fare_amount
False    286180
Name: count, dtype: int64

In [64]:
com_df.fare_amount.min()

0.0

Did you notice something different in the `RatecodeID` column for above records?

In [67]:
# Analyse RatecodeID for the negative fare amounts
com_df.RatecodeID.min()


1.0

In [65]:
com_df.RatecodeID.value_counts()

RatecodeID
1.0     260718
2.0      10774
5.0       1621
99.0      1595
3.0        943
4.0        644
6.0          1
Name: count, dtype: int64

In [66]:
com_df.RatecodeID.dtype

dtype('float64')

In [69]:
# Find which columns have negative values
com_df.PULocationID.min()
com_df.DOLocationID.min()



1

In [30]:
# fix these negative values



### **2.2** Handling Missing Values
<font color = red>[10 marks]</font> <br>

**2.2.1**  <font color = red>[2 marks]</font> <br>
Find the proportion of missing values in each column




In [71]:
missing_proportion = com_df.isnull().mean()
print("Proportion of missing values in each column:\n", missing_proportion)

Proportion of missing values in each column:
 VendorID                 0.000000
tpep_pickup_datetime     0.000000
tpep_dropoff_datetime    0.000000
passenger_count          0.034538
trip_distance            0.000000
RatecodeID               0.034538
store_and_fwd_flag       0.034538
PULocationID             0.000000
DOLocationID             0.000000
payment_type             0.000000
fare_amount              0.000000
extra                    0.000000
mta_tax                  0.000000
tip_amount               0.000000
tolls_amount             0.000000
improvement_surcharge    0.000000
total_amount             0.000000
congestion_surcharge     0.034538
Airport_fee              0.034538
dtype: float64


In [31]:
# Find the proportion of missing values in each column



**2.2.2**  <font color = red>[3 marks]</font> <br>
Handling missing values in `passenger_count`

In [72]:
# Display the rows with null values
# Impute NaN values in 'passenger_count'
com_df[com_df['passenger_count'].isnull()]

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
17             2  2023-02-25 14:13:23   2023-02-25 14:37:04              NaN   
26             2  2023-02-06 19:45:00   2023-02-06 19:58:00              NaN   
124            2  2023-02-20 19:58:28   2023-02-20 20:08:17              NaN   
192            2  2023-02-25 04:06:57   2023-02-25 04:35:10              NaN   
238            2  2023-02-01 08:50:00   2023-02-01 09:12:00              NaN   
...          ...                  ...                   ...              ...   
286101         2  2023-12-14 18:41:03   2023-12-14 18:54:52              NaN   
286128         2  2023-12-23 04:12:04   2023-12-23 04:19:38              NaN   
286142         2  2023-12-20 13:37:52   2023-12-20 13:49:49              NaN   
286145         2  2023-12-08 10:11:55   2023-12-08 10:28:18              NaN   
286146         2  2023-12-18 13:12:31   2023-12-18 13:16:44              NaN   

        trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
17               2.87         NaN               None           255   
26               5.20         NaN               None           140   
124              1.88         NaN               None           125   
192              8.12         NaN               None           158   
238              3.88         NaN               None           164   
...               ...         ...                ...           ...   
286101           1.48         NaN               None            68   
286128           1.83         NaN               None           186   
286142           1.04         NaN               None           237   
286145           1.15         NaN               None           143   
286146           0.76         NaN               None           107   

        DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
17                79             0        16.79    0.0      0.5        4.16   
26                79             0        14.76    0.0      0.5        3.75   
124              107             0        13.23    0.0      0.5        3.45   
192               85             0        33.23    0.0      0.5        7.45   
238              239             0        20.52    0.0      0.5        4.90   
...              ...           ...          ...    ...      ...         ...   
286101           100             0         7.90    0.0      0.5        3.00   
286128            48             0         8.18    0.0      0.5        0.00   
286142           237             0        11.58    0.0      0.5        1.56   
286145           230             0        10.88    0.0      0.5        0.00   
286146           224             0        26.83    0.0      0.5        0.00   

        tolls_amount  improvement_surcharge  total_amount  \
17               0.0                    1.0         24.95   
26               0.0                    1.0         22.51   
124              0.0                    1.0         20.68   
192              0.0                    1.0         44.68   
238              0.0                    1.0         29.42   
...              ...                    ...           ...   
286101           0.0                    1.0         14.90   
286128           0.0                    1.0         12.18   
286142           0.0                    1.0         17.14   
286145           0.0                    1.0         14.88   
286146           0.0                    1.0         30.83   

        congestion_surcharge  Airport_fee  
17                       NaN          NaN  
26                       NaN          NaN  
124                      NaN          NaN  
192                      NaN          NaN  
238                      NaN          NaN  
...                      ...          ...  
286101                   NaN          NaN  
286128                   NaN          NaN  
286142                   NaN          NaN  
286145                   NaN          NaN  
286146                   NaN          Na

Did you find zeroes in passenger_count? Handle these.

In [73]:
com_df.passenger_count.value_counts()

passenger_count
1.0    207950
2.0     41443
3.0     10564
4.0      5962
0.0      4328
5.0      3668
6.0      2378
8.0         3
Name: count, dtype: int64

In [77]:
com_df['passenger_count'] = com_df['passenger_count'].apply(lambda x: 1 if x == 0 else x)
com_df.passenger_count.value_counts()

passenger_count
1.0    212278
2.0     41443
3.0     10564
4.0      5962
5.0      3668
6.0      2378
8.0         3
Name: count, dtype: int64

**2.2.3**  <font color = red>[2 marks]</font> <br>
Handle missing values in `RatecodeID`

In [33]:
# Fix missing values in 'RatecodeID'


**2.2.4**  <font color = red>[3 marks]</font> <br>
Impute NaN in `congestion_surcharge`

In [34]:
# handle null values in congestion_surcharge


In [85]:
com_df['congestion_surcharge'] = com_df['congestion_surcharge'].apply(lambda x: 2.5 if x == -2.5 else x)
com_df.congestion_surcharge.value_counts()


congestion_surcharge
2.50    276295
0.75         1
Name: count, dtype: int64

In [87]:
com_df.improvement_surcharge.value_counts()


improvement_surcharge
 1.0    283001
-1.0      2903
 0.3       162
 0.0       112
-0.3         2
Name: count, dtype: int64

In [88]:
com_df['improvement_surcharge'] = com_df['improvement_surcharge'].apply(lambda x: 1 if x == -1.0 else x)
com_df['improvement_surcharge'] = com_df['improvement_surcharge'].apply(lambda x: 0.3 if x == -0.3 else x)
com_df['improvement_surcharge'] = com_df['improvement_surcharge'].apply(lambda x: 1 if x == 0 else x)
com_df.congestion_surcharge.value_counts()

congestion_surcharge
2.50    276295
0.75         1
Name: count, dtype: int64

Are there missing values in other columns? Did you find NaN values in some other set of columns? Handle those missing values below.

In [86]:
com_df.total_amount.min()

-393.5

In [35]:
# Handle any remaining missing values



### **2.3** Handling Outliers
<font color = red>[10 marks]</font> <br>

Before we start fixing outliers, let's perform outlier analysis.

In [36]:
# Describe the data and check if there are any potential outliers present
# Check for potential out of place values in various columns



**2.3.1**  <font color = red>[10 marks]</font> <br>
Based on the above analysis, it seems that some of the outliers are present due to errors in registering the trips. Fix the outliers.

Some points you can look for:
- Entries where `trip_distance` is nearly 0 and `fare_amount` is more than 300
- Entries where `trip_distance` and `fare_amount` are 0 but the pickup and dropoff zones are different (both distance and fare should not be zero for different zones)
- Entries where `trip_distance` is more than 250  miles.
- Entries where `payment_type` is 0 (there is no payment_type 0 defined in the data dictionary)

These are just some suggestions. You can handle outliers in any way you wish, using the insights from above outlier analysis.

How will you fix each of these values? Which ones will you drop and which ones will you replace?

First, let us remove 7+ passenger counts as there are very less instances.

In [37]:
# remove passenger_count > 6


In [38]:
# Continue with outlier handling



In [39]:
# Do any columns need standardising?



## **3** Exploratory Data Analysis
<font color = red>[90 marks]</font> <br>

In [ ]:
df.columns.tolist()

#### **3.1** General EDA: Finding Patterns and Trends
<font color = red>[40 marks]</font> <br>

**3.1.1** <font color = red>[3 marks]</font> <br>
Categorise the varaibles into Numerical or Categorical.
* `VendorID`:
* `tpep_pickup_datetime`:
* `tpep_dropoff_datetime`:
* `passenger_count`:
* `trip_distance`:
* `RatecodeID`:
* `PULocationID`:
* `DOLocationID`:
* `payment_type`:
* `pickup_hour`:
* `trip_duration`:


The following monetary parameters belong in the same category, is it categorical or numerical?


* `fare_amount`
* `extra`
* `mta_tax`
* `tip_amount`
* `tolls_amount`
* `improvement_surcharge`
* `total_amount`
* `congestion_surcharge`
* `airport_fee`

##### Temporal Analysis

**3.1.2** <font color = red>[5 marks]</font> <br>
Analyse the distribution of taxi pickups by hours, days of the week, and months.

In [ ]:
# Find and show the hourly trends in taxi pickups



In [ ]:
# Find and show the daily trends in taxi pickups (days of the week)



In [ ]:
# Show the monthly trends in pickups



##### Financial Analysis

Take a look at the financial parameters like `fare_amount`, `tip_amount`, `total_amount`, and also `trip_distance`. Do these contain zero/negative values?

In [ ]:
# Analyse the above parameters



Do you think it is beneficial to create a copy DataFrame leaving out the zero values from these?

**3.1.3** <font color = red>[2 marks]</font> <br>
Filter out the zero values from the above columns.

**Note:** The distance might be 0 in cases where pickup and drop is in the same zone. Do you think it is suitable to drop such cases of zero distance?

In [ ]:
# Create a df with non zero entries for the selected parameters.



**3.1.4** <font color = red>[3 marks]</font> <br>
Analyse the monthly revenue (`total_amount`) trend

In [ ]:
# Group data by month and analyse monthly revenue



**3.1.5** <font color = red>[3 marks]</font> <br>
Show the proportion of each quarter of the year in the revenue

In [ ]:
# Calculate proportion of each quarter



**3.1.6** <font color = red>[3 marks]</font> <br>
Visualise the relationship between `trip_distance` and `fare_amount`. Also find the correlation value for these two.

**Hint:** You can leave out the trips with trip_distance = 0

In [ ]:
# Show how trip fare is affected by distance



**3.1.7** <font color = red>[5 marks]</font> <br>
Find and visualise the correlation between:
1. `fare_amount` and trip duration (pickup time to dropoff time)
2. `fare_amount` and `passenger_count`
3. `tip_amount` and `trip_distance`

In [ ]:
# Show relationship between fare and trip duration



In [ ]:
# Show relationship between fare and number of passengers



In [ ]:
# Show relationship between tip and trip distance



**3.1.8** <font color = red>[3 marks]</font> <br>
Analyse the distribution of different payment types (`payment_type`)

In [ ]:
# Analyse the distribution of different payment types (payment_type).




- 1= Credit card
- 2= Cash
- 3= No charge
- 4= Dispute



##### Geographical Analysis

For this, you have to use the *taxi_zones.shp* file from the *taxi_zones* folder.

There would be multiple files inside the folder (such as *.shx, .sbx, .sbn* etc). You do not need to import/read any of the files other than the shapefile, *taxi_zones.shp*.

Do not change any folder structure - all the files need to be present inside the folder for it to work.

The folder structure should look like this:
```
Taxi Zones
|- taxi_zones.shp.xml
|- taxi_zones.prj
|- taxi_zones.sbn
|- taxi_zones.shp
|- taxi_zones.dbf
|- taxi_zones.shx
|- taxi_zones.sbx

 ```

 You only need to read the `taxi_zones.shp` file. The *shp* file will utilise the other files by itself.

We will use the *GeoPandas* library for geopgraphical analysis
```
import geopandas as gpd
```

More about geopandas and shapefiles: [About](https://geopandas.org/en/stable/about.html)


Reading the shapefile is very similar to *Pandas*. Use `gpd.read_file()` function to load the data (*taxi_zones.shp*) as a GeoDataFrame. Documentation: [Reading and Writing Files](https://geopandas.org/en/stable/docs/user_guide/io.html)

In [ ]:
# !pip install geopandas

**3.1.9** <font color = red>[2 marks]</font> <br>
Load the shapefile and display it.

In [ ]:
# import geopandas as gpd


# Read the shapefile using geopandas
zones = # read the .shp file using gpd
zones.head()

Now, if you look at the DataFrame created, you will see columns like: `OBJECTID`,`Shape_Leng`, `Shape_Area`, `zone`, `LocationID`, `borough`, `geometry`.
<br><br>

Now, the `locationID` here is also what we are using to mark pickup and drop zones in the trip records.

The geometric parameters like shape length, shape area and geometry are used to plot the zones on a map.

This can be easily done using the `plot()` method.

In [ ]:
# print(zones.info())
# zones.plot()

Now, you have to merge the trip records and zones data using the location IDs.



**3.1.10** <font color = red>[3 marks]</font> <br>
Merge the zones data into trip data using the `locationID` and `PULocationID` columns.

In [ ]:
# Merge zones and trip records using locationID and PULocationID



**3.1.11** <font color = red>[3 marks]</font> <br>
Group data by location IDs to find the total number of trips per location ID

In [ ]:
# Group data by location and calculate the number of trips



**3.1.12** <font color = red>[2 marks]</font> <br>
Now, use the grouped data to add number of trips to the GeoDataFrame.

We will use this to plot a map of zones showing total trips per zone.

In [ ]:
# Merge trip counts back to the zones GeoDataFrame




The next step is creating a color map (choropleth map) showing zones by the number of trips taken.

Again, you can use the `zones.plot()` method for this. [Plot Method GPD](https://geopandas.org/en/stable/docs/reference/api/geopandas.GeoDataFrame.plot.html#geopandas.GeoDataFrame.plot)

But first, you need to define the figure and axis for the plot.

`fig, ax = plt.subplots(1, 1, figsize = (12, 10))`

This function creates a figure (fig) and a single subplot (ax)

---

After setting up the figure and axis, we can proceed to plot the GeoDataFrame on this axis. This is done in the next step where we use the plot method of the GeoDataFrame.

You can define the following parameters in the `zones.plot()` method:
```
column = '',
ax = ax,
legend = True,
legend_kwds = {'label': "label", 'orientation': "<horizontal/vertical>"}
```

To display the plot, use `plt.show()`.

**3.1.13** <font color = red>[3 marks]</font> <br>
Plot a color-coded map showing zone-wise trips

In [ ]:
# Define figure and axis


# Plot the map and display it



In [ ]:
# can you try displaying the zones DF sorted by the number of trips?



Here we have completed the temporal, financial and geographical analysis on the trip records.

**Compile your findings from general analysis below:**

You can consider the following points:

* Busiest hours, days and months
* Trends in revenue collected
* Trends in quarterly revenue
* How fare depends on trip distance, trip duration and passenger counts
* How tip amount depends on trip distance
* Busiest zones


#### **3.2** Detailed EDA: Insights and Strategies
<font color = red>[50 marks]</font> <br>

Having performed basic analyses for finding trends and patterns, we will now move on to some detailed analysis focussed on operational efficiency, pricing strategies, and customer experience.

##### Operational Efficiency

Analyze variations by time of day and location to identify bottlenecks or inefficiencies in routes

**3.2.1** <font color = red>[3 marks]</font> <br>
Identify slow routes by calculating the average time taken by cabs to get from one zone to another at different hours of the day.

Speed on a route *X* for hour *Y* = (*distance of the route X / average trip duration for hour Y*)

In [ ]:
# Find routes which have the slowest speeds at different times of the day



How does identifying high-traffic, high-demand routes help us?

**3.2.2** <font color = red>[3 marks]</font> <br>
Calculate the number of trips at each hour of the day and visualise them. Find the busiest hour and show the number of trips for that hour.

In [ ]:
# Visualise the number of trips per hour and find the busiest hour



Remember, we took a fraction of trips. To find the actual number, you have to scale the number up by the sampling ratio.

**3.2.3** <font color = red>[2 mark]</font> <br>
Find the actual number of trips in the five busiest hours

In [ ]:
# Scale up the number of trips

# Fill in the value of your sampling fraction and use that to scale up the numbers
sample_fraction =



**3.2.4** <font color = red>[3 marks]</font> <br>
Compare hourly traffic pattern on weekdays. Also compare for weekend.

In [ ]:
# Compare traffic trends for the week days and weekends



What can you infer from the above patterns? How will finding busy and quiet hours for each day help us?

**3.2.5** <font color = red>[3 marks]</font> <br>
Identify top 10 zones with high hourly pickups. Do the same for hourly dropoffs. Show pickup and dropoff trends in these zones.

In [ ]:
# Find top 10 pickup and dropoff zones



**3.2.6** <font color = red>[3 marks]</font> <br>
Find the ratio of pickups and dropoffs in each zone. Display the 10 highest (pickup/drop) and 10 lowest (pickup/drop) ratios.

In [ ]:
# Find the top 10 and bottom 10 pickup/dropoff ratios



**3.2.7** <font color = red>[3 marks]</font> <br>
Identify zones with high pickup and dropoff traffic during night hours (11PM to 5AM)

In [ ]:
# During night hours (11pm to 5am) find the top 10 pickup and dropoff zones
# Note that the top zones should be of night hours and not the overall top zones



Now, let us find the revenue share for the night time hours and the day time hours. After this, we will move to deciding a pricing strategy.

**3.2.8** <font color = red>[2 marks]</font> <br>
Find the revenue share for nighttime and daytime hours.

In [ ]:
# Filter for night hours (11 PM to 5 AM)



##### Pricing Strategy

**3.2.9** <font color = red>[2 marks]</font> <br>
For the different passenger counts, find the average fare per mile per passenger.

For instance, suppose the average fare per mile for trips with 3 passengers is 3 USD/mile, then the fare per mile per passenger will be 1 USD/mile.

In [ ]:
# Analyse the fare per mile per passenger for different passenger counts




**3.2.10** <font color = red>[3 marks]</font> <br>
Find the average fare per mile by hours of the day and by days of the week

In [ ]:
# Compare the average fare per mile for different days and for different times of the day



**3.2.11** <font color = red>[3 marks]</font> <br>
Analyse the average fare per mile for the different vendors for different hours of the day

In [ ]:
# Compare fare per mile for different vendors



**3.2.12** <font color = red>[5 marks]</font> <br>
Compare the fare rates of the different vendors in a tiered fashion. Analyse the average fare per mile for distances upto 2 miles. Analyse the fare per mile for distances from 2 to 5 miles. And then for distances more than 5 miles.


In [ ]:
# Defining distance tiers



##### Customer Experience and Other Factors

**3.2.13** <font color = red>[5 marks]</font> <br>
Analyse average tip percentages based on trip distances, passenger counts and time of pickup. What factors lead to low tip percentages?

In [ ]:
#  Analyze tip percentages based on distances, passenger counts and pickup times



Additional analysis [optional]: Let's try comparing cases of low tips with cases of high tips to find out if we find a clear aspect that drives up the tipping behaviours

In [ ]:
# Compare trips with tip percentage < 10% to trips with tip percentage > 25%



**3.2.14** <font color = red>[3 marks]</font> <br>
Analyse the variation of passenger count across hours and days of the week.

In [ ]:
# See how passenger count varies across hours and days




**3.2.15** <font color = red>[2 marks]</font> <br>
Analyse the variation of passenger counts across zones

In [ ]:
# How does passenger count vary across zones



In [ ]:
# For a more detailed analysis, we can use the zones_with_trips GeoDataFrame
# Create a new column for the average passenger count in each zone.



Find out how often surcharges/extra charges are applied to understand their prevalance

**3.2.16** <font color = red>[5 marks]</font> <br>
Analyse the pickup/dropoff zones or times when extra charges are applied more frequently

In [ ]:
# How often is each surcharge applied?



## **4** Conclusion
<font color = red>[15 marks]</font> <br>

### **4.1** Final Insights and Recommendations
<font color = red>[15 marks]</font> <br>

Conclude your analyses here. Include all the outcomes you found based on the analysis.

Based on the insights, frame a concluding story explaining suitable parameters such as location, time of the day, day of the week etc. to be kept in mind while devising a strategy to meet customer demand and optimise supply.

**4.1.1** <font color = red>[5 marks]</font> <br>
Recommendations to optimize routing and dispatching based on demand patterns and operational inefficiencies

**4.1.2** <font color = red>[5 marks]</font> <br>

Suggestions on strategically positioning cabs across different zones to make best use of insights uncovered by analysing trip trends across time, days and months.

**4.1.3** <font color = red>[5 marks]</font> <br>
Propose data-driven adjustments to the pricing strategy to maximize revenue while maintaining competitive rates with other vendors.